sdks/go/pkg/beam/core/graph/coder/testutil/testutil_test.go

```diff
@@ -0,0 +1,126 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package testutil
+
+import (
+	"fmt"
+	"io"
+	"reflect"
+	"testing"
+
+	"github.com/apache/beam/sdks/go/pkg/beam/core/graph/coder"
+)
+
+type UserInterface interface {
+	mark()
+}
+
+type UserType1 struct {
+	A string
+	B int
+	C string
+}
+
+func (UserType1) mark() {}
+
+func ut1EncDropB(val interface{}, w io.Writer) error {
+	if err := coder.WriteSimpleRowHeader(2, w); err != nil {
+		return err
+	}
+	elm := val.(UserType1)
+	if err := coder.EncodeStringUTF8(elm.A, w); err != nil {
+		return err
+	}
+	if err := coder.EncodeStringUTF8(elm.C, w); err != nil {
+		return err
+	}
+	return nil
+}
+
+func ut1DecDropB(r io.Reader) (interface{}, error) {
+	if err := coder.ReadSimpleRowHeader(2, r); err != nil {
+		return nil, err
+	}
+	a, err := coder.DecodeStringUTF8(r)
+	if err != nil {
+		return nil, fmt.Errorf("decoding string field A: %w", err)
+	}
+	c, err := coder.DecodeStringUTF8(r)
+	if err != nil {
+		return nil, fmt.Errorf("decoding string field C: %v, %w", c, err)
+	}
+	return UserType1{
+		A: a,
+		B: 42,
+		C: c,
+	}, nil
+}
+
+// TestValidateCoder_SingleValue checks that the validate coder fun will
+func TestValidateCoder(t *testing.T) {
+	// Validates a custom UserType1 encoding, which drops encoding the "B" field,
+	// always setting it to a constant value.
+	t.Run("SingleValue", func(t *testing.T) {
+		(&SchemaCoder{}).Validate(t, reflect.TypeOf((*UserType1)(nil)).Elem(),
+			func(reflect.Type) (func(interface{}, io.Writer) error, error) { return ut1EncDropB, nil },
+			func(reflect.Type) (func(io.Reader) (interface{}, error), error) { return ut1DecDropB, nil },
+			struct{ A, C string }{},
+			UserType1{
+				A: "cats",
+				B: 42,
+				C: "pjamas",
+			},
+		)
+	})
+	t.Run("SliceOfValues", func(t *testing.T) {
+		(&SchemaCoder{}).Validate(t, reflect.TypeOf((*UserType1)(nil)).Elem(),
+			func(reflect.Type) (func(interface{}, io.Writer) error, error) { return ut1EncDropB, nil },
+			func(reflect.Type) (func(io.Reader) (interface{}, error), error) { return ut1DecDropB, nil },
+			struct{ A, C string }{},
+			[]UserType1{
+				{
+					A: "cats",
+					B: 42,
+					C: "pjamas",
+				}, {
+					A: "dogs",
+					B: 42,
+					C: "breakfast",
+				}, {
+					A: "fish",
+					B: 42,
+					C: "plenty of",
+				},
+			},
+		)
+	})
+	t.Run("InterfaceCoder", func(t *testing.T) {
+		(&SchemaCoder{}).Validate(t, reflect.TypeOf((*UserInterface)(nil)).Elem(),
+			func(rt reflect.Type) (func(interface{}, io.Writer) error, error) {
+				return ut1EncDropB, nil
+			},
+			func(rt reflect.Type) (func(io.Reader) (interface{}, error), error) {
+				return ut1DecDropB, nil
+			},
+			struct{ A, C string }{},
+			UserType1{
+				A: "cats",
+				B: 42,
+				C: "pjamas",
+			},
+		)
+	})
+}
```

Comment:
Suggestion: You could probably add a test for the failure case though you'll need to do some refactoring on the Validate function. Basically just have an inner function that's the current implementation but doesn't use `*testing.T` and instead returns an error if something went wrong and make the current exported `Validate` function call that and do `t.Fatalf` if an error appears. With that refactor, you can add a test that calls the inner function with an element that doesn't have 42 for `B`, so you can validate that it fails when expected.

Reply:
Fair, I could swap *testing.T with an interface with all the common methods I use. 
I can't do anything about the interior coder tests though. Part of the value of this set up is that the coder is tested multiple times with whatever different values are in test values, but they fail independently. 

I can check the first three fail as expected though.

I ran into the various error cases on my implementation of this so I'm not that worried. That's part of why I'm not 100% on the documentation+API just yet. This will be refined as I work through the proto provider.